components/toast.js

```diff
@@ -0,0 +1,51 @@
+import React, { createContext, useCallback, useContext, useState } from 'react'
+import Toast from 'react-bootstrap/Toast'
+import ToastBody from 'react-bootstrap/ToastBody'
+import ToastContainer from 'react-bootstrap/ToastContainer'
+import ToastHeader from 'react-bootstrap/ToastHeader'
+import { randInRange } from '../lib/rand'
+
+const ToastContext = createContext(() => {})
+
+export const ToastProvider = ({ children }) => {
+  const [toasts, setToasts] = useState([])
+  const dispatchToast = useCallback((toastConfig) => {
+    const id = Array(10).fill(randInRange(0, 9)).join('')
```

Comment:
You just need an id that is random enough for no collisions, right?

`randInRange` returns float and the value in `fill` is just executed once so the resulting id looks kind of weird:

```
$ node
Welcome to Node.js v20.5.0.
Type ".help" for more information.
> const randInRange = (min, max) => Math.random() * (max - min) + min
undefined
> randInRange(0, 10)
1.13396597332184
> randInRange(0, 10)
1.2429912934983878
> randInRange(0, 10)
3.6246351810399124
> Array(10).fill(randInRange(0, 9)).join('')
'0.482431026012052340.482431026012052340.482431026012052340.482431026012052340.482431026012052340.482431026012052340.482431026012052340.482431026012052340.482431026012052340.48243102601205234'
```

Can't you just use `toasts.length` as the id? I assume the id just has to be unique among all _existing_ toasts

(You then also need to add `toasts` as a dependeny of `useCallback`)

Reply:
Yes, I do just need it to be unique to avoid collisions. I did the random code in lieu of access to a `uuid` generator function, I didn't want to add a dependency for it.

I don't think using `toasts.length` is safe enough. Consider the following:

1. dispatch a sticky toast -> id: 0. `toasts` looks like: `[{ id: 0}]`
2. dispatch an auto-dismissed toast: id: 1. `toasts` looks like: `[{ id: 0}, {id: 1}]`
3. user dismisses toast 0 (makes toasts' length 1). `toasts` looks like: `[{id: 1}]`
4. dispatch an auto-dismissed toast: id: 1. `toasts` looks like: `[{id: 1}, {id: 1}]` -> duplicate id

Admittedly, I was thinking `randInRange` would give an integer, not a float.

I think I'll just refactor to use a globally incrementing integer, starting at 0. If a user has `Number.MAX_SAFE_INTEGER` toasts in one page session, they can afford to refresh the page ;)

Reply:
> I don't think using toasts.length is safe enough. Consider the following:

Ah, makes sense. I thought for some reason the IDs would always correspond to the index.

> I think I'll just refactor to use a globally incrementing integer, starting at 0. If a user has Number.MAX_SAFE_INTEGER toasts in one page session, they can afford to refresh the page ;)

Yes, globally incrementing integer makes sense.